doc/development/tutorials/extending_syntax.rst

```diff
@@ -0,0 +1,224 @@
+
+With this we will add the extension to the project,
+within the :file:`source` folder:
+
+#. Create an :file:`_ext` folder in :file:`source`
+#. Create a new Python file in the :file:`_ext` folder called
+   :file:`helloworld.py`
+
+Here is an example of the folder structure you might obtain:
+
+.. code-block:: text
+
+      └── source
+          ├── _ext
+          │   └── helloworld.py
+          ├── conf.py
+          ├── index.rst
+
+
+Writing the extension
+---------------------
+
+Open :file:`helloworld.py` and paste the following code in it:
+
+.. literalinclude:: examples/helloworld.py
+   :language: python
+   :linenos:
+
+Some essential things are happening in this example:
+
+The role class
+...............
+
+Our new role is declared in the ``HelloRole`` class.
+
+.. literalinclude:: examples/helloworld.py
+   :language: python
+   :linenos:
+   :lines: 10-15
+
+This class extends the :class:`.SphinxRole` class.
+The class contains a ``run`` method,
+which is a requirement for every role.
+It contains the main logic of the role and it
+returns a tuple containing:
+
+- a list of inline-level docutils nodes to be processed by Sphinx.
+- an (optional) list of system message nodes
+
+The directive class
+...................
+
+Our new directive is declared in the ``HelloDirective`` class.
+
+.. literalinclude:: examples/helloworld.py
+   :language: python
+   :linenos:
+   :lines: 18-25
+
+This class extends the :class:`.SphinxDirective` class.
+The class contains a ``run`` method,
+which is a requirement for every directive.
+It contains the main logic of the directive and it
+returns a list of block-level docutils nodes to be processed by Sphinx.
+It also contains a ``required_arguments`` attribute,
+which tells Sphinx how many arguments are required for the directive.
+
+What are docutils nodes?
+........................
+
+When Sphinx parses a document,
+it creates an "Abstract Syntax Tree" (AST) of nodes
+that represent the content of the document in a structured way,
+that is generally independent of any one
+input (rST, MyST, etc) or output (HTML, LaTeX, etc) format.
+It is a tree because each node can have children nodes, and so on:
+
+.. code-block:: xml
+
+      <document>
+         <paragraph>
+            <text>
+               Hallo world!
+
+The docutils_ package provides many `built-in nodes <docutils nodes_>`_,
+to represent different types of content such as
+text, paragraphs, references, tables, etc.
+
+Each node type generally only accepts a specific set of direct child nodes,
+for example the ``document`` node should only contain "block-level" nodes,
+such as ``paragraph``, ``section``, ``table``, etc,
+whilst the ``paragraph`` node should only contain "inline-level" nodes,
+such as ``text``, ``emphasis``, ``strong``, etc.
+
+.. seealso::
+
+   The docutils documentation on
+   `creating directives <docutils directives_>`_, and
+   `creating roles <docutils roles_>`_.
+
+The ``setup`` function
```

Comment:
Slightly confusing terminology/naming emphasis here, because there is no object/argument named `setup` that is used (AFAICT?)

Reply:
(maybe I'm confused - it does seem that extensions have to have a `setup` function.  is that also true for directives?)

Reply:
oh yes, an extension always has to have a `setup` function; that's the only way you load things into a sphinx build (including directives)

Reply:
Ok, got it - thanks!  I think I'd suggest updating the line reference from `30-` to `28-` to include the method signature in the snippet shown alongside.

Reply:
Nope, I have a better suggestion.. hold on..

Reply:
ah well just changed it to `28-` (that was a mistake by me, it was meant to show all of `setup`)